scripts/zowe-install-extension.sh

```diff
@@ -0,0 +1,148 @@
+#!/bin/sh
+
+SCRIPT_NAME="$(basename $0)"
+EXTENSION_PATH=""
+INSTANCE_DIR=""
+TARGET_DIR=""
+EXTENSION_NAME=""
+INSTALL_TYPE=""
+
+error_handler(){
+    echo "$SCRIPT_NAME: $1" 1>&2
+    exit 1
+}
+
+install_component(){
+    cd $TARGET_DIR
+
+    if [ ! -d "$EXTENSION_NAME" ]; then
+        mkdir $EXTENSION_NAME
+    else
+        # Clean up the directory and make a new one
+        rm -rf $EXTENSION_NAME
+        mkdir $EXTENSION_NAME
+    fi
+
+    cd $EXTENSION_NAME
+
+    pax -ppx -rf "$EXTENSION_PATH"
+
+    npm install --only=prod
+
+    # Validates the pax file is correctly configured when unpaxed
+    if [ ! -d "$TARGET_DIR/$EXTENSION_NAME/bin" ] || [ ! -e "$TARGET_DIR/$EXTENSION_NAME/bin/start.sh" ]; then
+        #error_handler "PAX file is not a zowe extension."
+        error_handler "$EXTENSION_PATH: Invalid package"
+    fi
+
+    # Represents the current value of EXTERNAL_COMPONENT variable in instance.env
+    EXT_COMP_VAL=$(eval "grep 'EXTERNAL_COMPONENTS' $INSTANCE_DIR/instance.env | cut -f2 -d= | cut -f1 -d#")
+
+    # Must add backslash escapees or else sed command will not work in the next step
+    EXTENSION_BIN_PATH=$(eval "echo "$TARGET_DIR/$EXTENSION_NAME/bin" | sed 's/\//\\\\\//g'")
+
+    if [ "$EXT_COMP_VAL" = " " ]; then
+        sed -e "s/EXTERNAL_COMPONENTS=/EXTERNAL_COMPONENTS=$EXTENSION_BIN_PATH/" $INSTANCE_DIR/instance.env > $INSTANCE_DIR/instance.env.tmp
+        mv $INSTANCE_DIR/instance.env.tmp $INSTANCE_DIR/instance.env
+    else
+        # Ensures that the bin directory of the extension is included into the instance.env once (Avoids duplication if same extension is installed twice)
+        if [[ $(grep 'EXTERNAL_COMPONENT' $INSTANCE_DIR/instance.env | grep $TARGET_DIR/$EXTENSION_NAME/bin) = "" ]]; then
+            sed -e "s/EXTERNAL_COMPONENTS=/EXTERNAL_COMPONENTS=$EXTENSION_BIN_PATH,/" $INSTANCE_DIR/instance.env > $INSTANCE_DIR/instance.env.tmp
+            mv $INSTANCE_DIR/instance.env.tmp $INSTANCE_DIR/instance.env
+        fi
+    fi
+
+    echo "Zowe component has been installed."
+}
+
+install_desktop_plugin(){
+    cd $TARGET_DIR
+
+    if [ ! -d "$EXTENSION_NAME" ]; then
+        mkdir $EXTENSION_NAME
+    else
+        # Clean up the directory and make a new one
+        rm -rf $EXTENSION_NAME
+        mkdir $EXTENSION_NAME
+    fi
+
+    cd $EXTENSION_NAME
+
+    pax -ppx -rf "$EXTENSION_PATH"
+
+    # Validates the pax file is correctly configured when unpaxed
+    if [ ! -e "pluginDefinition.json" ]; then
+        error_handler "$EXTENSION_PATH: Invalid package"
+    fi
+
+    # Uses install-app.sh in zowe-instance-dir to automatically set up the extension onto zowe
+    $INSTANCE_DIR/bin/install-app.sh $TARGET_DIR/$EXTENSION_NAME
+
+    echo "Zowe desktop plugin has been installed."
+}
+
```

Comment:
please use /bin/internal/zowe-set-env.sh to ensure that common zowe script environment properties are set